stdlib/public/core/Diffing.swift

```diff
@@ -192,615 +165,168 @@ extension BidirectionalCollection where Element : Equatable {
-private struct _CountingIndexCollection<Base: BidirectionalCollection> {
-  private let base: Base
-
-  fileprivate init(_ base: Base) {
-    self.base = base
-  }
-}
-
-extension _CountingIndexCollection : BidirectionalCollection {
-  fileprivate typealias Index = _CountingIndex<Base.Index>
-  fileprivate typealias Element = Base.Element
-
-  fileprivate var startIndex: Index {
-    return Index(base: base.startIndex, offset: base.isEmpty ? nil : 0)
-  }
-
-  fileprivate var endIndex: Index {
-    return Index(base: base.endIndex, offset: nil)
-  }
-
-  fileprivate func index(after i: Index) -> Index {
-    let next = base.index(after: i.base)
-    return Index(
-      base: next, offset: next == base.endIndex ? nil : i.offset! + 1)
-  }
-
-  fileprivate func index(before i: Index) -> Index {
-    let prev = base.index(before: i.base)
-    return Index(
-      base: prev, offset: prev == base.endIndex ? nil : i.offset! + 1)
-  }
+    assert(x >= n && y >= m)
 
-  fileprivate subscript(position: Index) -> Element {
-    return base[position.base]
+    return result
   }
-}
 
-/// Returns the nth [triangular number].
-///
-/// [triangular number]: https://en.wikipedia.org/wiki/Triangular_number
-fileprivate func _triangularNumber(_ n: Int) -> Int {
-  return n * (n + 1) / 2
-}
-
-/// A square matrix that only provides subscript access to elements on, or
-/// below, the main diagonal.
-///
-/// A [lower triangular matrix] can be dynamically grown:
-///
-///   var m = _LowerTriangularMatrix<Int>()
-///   m.appendRow(repeating: 1)
-///   m.appendRow(repeating: 2)
-///   m.appendRow(repeating: 3)
-///
-///   assert(Array(m.rowMajorOrder) == [
-///     1,
-///     2, 2,
-///     3, 3, 3,
-///   ])
-///
-/// [lower triangular matrix]: http://en.wikipedia.org/wiki/Triangular_matrix
-fileprivate struct _LowerTriangularMatrix<Element> {
-  /// The matrix elements stored in [row major order][rmo].
-  ///
-  /// [rmo]: http://en.wikipedia.org/wiki/Row-_and_column-major_order
-  fileprivate var storage: [Element] = []
-
-  /// The dimension of the matrix.
-  ///
-  /// Being a square matrix, the number of rows and columns are equal.
-  fileprivate var dimension: Int = 0
-
-  fileprivate subscript(row: Int, column: Int) -> Element {
-    get {
-      _internalInvariant((0...row).contains(column))
-      return storage[_triangularNumber(row) + column]
+  /* Splatting the collections into arrays here has two advantages:
+   *
+   *   1) Subscript access becomes inlined
+   *   2) Subscript index becomes Int, matching the iterator types in the algorithm
+   *
+   * Combined, these effects dramatically improves performance when
+   * collections differ significantly, without unduly degrading runtime when
+   * the parameters are very similar.
+   *
+   * In terms of memory use, the linear cost is significantly less than the
+   * worst-case n² memory use of the descent algorithm.
+   */
+  let a = Array(old)
+  let b = Array(new)
```

Comment:
I hear that these performance improvements may be possible to accomplish at a lower cost using `withContiguousStorageIfAvailable`?

Reply:
withContiguousStorageIfAvailable should, if contiguous storage is in fact available, be decently faster in fact. Array() will allocate

Reply:
Ok, we need to measure that the result of `withContiguousStorage` is still as fast as using a concrete `Array` before adopting it—the cost of `Array()` barely showed up in profiles, and the savings in removing the indirection on subscript is massive 🙃

Reply:
We can still have the indirection removed by having the actual code work on UnsafeBufferPointers instead of arrays. I.e., if withContiguousStorage works, then use the buffer provided, else create the arrays and work with the buffers from those.

The weird optionality of the withContiguousStorage API makes this a bit tricky, but it's possible:

```swift
// untested
private func _withContiguousStorage<C: Collection, R>(for values: C, _ body: (UBP<Element>) -> R) rethrows -> R {
  if let result = try values.withContiguousStorageIfAvailable(body) { return result }
  let array = Array(values)
  return try array.withUnsafeBufferPointer(body)
}

...

return _withContiguousStorage(for: old) { a in 
  _withContiguousStorage(for: new) { b in
    // diffing magic
  }
}
```

Switching to UnsafeBufferPointer does have the drawback of disabling range checks for subscripts in release builds of the stdlib. That should be a slight performance boost, but obviously it can cause significant headaches if there's a bug.

Reply:
We can still have the indirection removed by having the actual code work on UnsafeBufferPointers instead of arrays. I.e., if withContiguousStorage works, then use the buffer provided, else create the arrays and work with the buffers from those.

The weird optionality of the withContiguousStorage API makes this a bit tricky, but it's possible:

```swift
// untested
private func _withContiguousStorage<C: Collection, R>(for values: C, _ body: (UBP<Element>) -> R) rethrows -> R {
  if let result = try values.withContiguousStorageIfAvailable(body) { return result }
  let array = Array(values)
  return try array.withUnsafeBufferPointer(body)
}

...

return _withContiguousStorage(for: old) { a in 
  _withContiguousStorage(for: new) { b in
    // diffing magic
  }
}
```

Switching to UnsafeBufferPointer does have the drawback of disabling range checks for subscripts in release builds of the stdlib. That should be a slight performance boost, but obviously it can cause significant headaches if there's a bug.